handle/Timer.lua

```diff
@@ -1,30 +1,190 @@
 require('CeresStdLib.base.Native')
 require('CeresStdLib.base.Log')
-local free_timers = {}
+
+local Timer 		= {}
+local free_timers	= {}
+local MAX_PUSHES	= 10000		--	Maximum amount of times push is called before garbage collection.
+
+free_timers[0]		= 0
+free_timers[-1]		= 0
+
+free_timers.has = function()
+	if free_timers[0] <= 0 then return false end return true
+end
+
+free_timers.push = function(t)
+	if free_timers[-1] >= MAX_PUSHES then
+		collectgarbage()
+	end
+	if not free_timers[t] then
+		free_timers[0] 					= free_timers[0] + 1
+		free_timers[t] 					= true
+		free_timers[free_timers[0]]		= t
+		free_timers[-1]					= free_timers[-1] + 1
+		return true
+	end
+	return false
+end
+
+free_timers.pop = function()
+	local obj
+	if free_timers[0] > 0 then
+		obj = free_timers[free_timers[0]]
+		free_timers[obj] 	= nil
+		free_timers[0] 		= free_timers[0] - 1
+		return obj
+	end
+	return nil
+end
 
 replaceNative('CreateTimer', function()
-	k, v = next(free_timers)
-	if k ~= nil then
-		free_timers[k] = nil
-		return k
+	local obj
+	if free_timers.has() then
+		obj = free_timers.pop()
+	else
+		obj = Native.CreateTimer()
+	end
+	Timer[obj]      = {
+		hasCallback = nil,
+		callback    = nil,
+		looped      = nil,
+		running     = nil,
+		inCallback  = nil,
+		pauseFlag   = 0,
+		duration    = 0.,
+		elapsed     = 0.
+	}
+	return obj
+end)
+
+replaceNative('PauseTimer', function(t)
+	if Timer[t].running then
+		Timer[t].running    = nil
+		Timer[t].pauseFlag  = (BlzBitOr(Timer[t].pauseFlag, 1))
+	end
+	Native.PauseTimer(t)
+end)
+
+replaceNative('ResumeTimer', function(t)
+	if not Timer[t].hasCallback then
+		Log.error("ResumeTimer: Attempted to resume a timer " .. I2S(GetHandleId(t)) .. " with no callback")
+		return
+	end
+	if Timer[t].inCallback and Timer[t].looped then
+		return
+	end
+	if not Timer[t].inCallback and not Timer[t].running then
+		Timer[t].running    = 0
+	end
+	Native.ResumeTimer(t)
+end)
+
+replaceNative('TimerStart', function(t, dur, looper, func)
+	if not Timer[t].hasCallback then
+		--  Make it come true
+		Timer[t].hasCallback = 0
```

Comment:
Then what's the point of the check at all? Remove both if closures and leave only this assignment

Reply:
Removed the first if condition.